File: 2017/Всесибирская/Закл 2015/4/4.py
fin = open('input.txt', 'r')
fout = open('output.txt', 'w')

n = int(fin.readline().strip())
#n, d = map(int, fin.readline().strip().split())
#num = ['']*(n+2)
#for i in range (1, n+1):
#   num[i] = fin.readline().strip()
el = []
ss = ''
u = [0]*12
ans = []
s = ''

def check():
    #print('CHECK\n')
    ch = 0
    s = ''
    for i in range (0, n):
        s += num[int(ss[i])]
    for i in range (0, len(s)):
        ch += int(s[i])*pow(10, len(s)-i-1)
    if (not(ch%d)): return True
    return False

def dfs_R(ind):
    global ss
    #print('DFS : ', ind)
    if (ind == n): #and check()):
        ans.append(ss)
    else:
        for i in range (1, n+1):
            if not(u[i]):
                u[i] = 1
                ss = ss[:ind]+str(i)+ss[ind+1:]
                dfs_R(ind+1)
                u[i] = 0
    
for i in range(1, n+1):
    el.append(i)
    ss += ' '
dfs_R(0)
#print(len(ans), file = fout)
ans.sort()
for i in ans:
    print(i)
fout.close()    

File: 2017/Всесибирская/Закл 2015/4/4.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <string>
#include <cmath>
#include <vector>
#include <iomanip>
#include <set>
#include <map>
#include <utility>
#include <queue>
#include <string.h>

#define SZ size()
#define VEC vector
#define PB push_back
#define MP make_pair
#define SS second
#define FF first
#define PII pair <int,int>
#define LL long long

using namespace std;

int used[10];

int n, d;
string s, str;
VEC <int> el;
VEC <string> num;
set <string> ans;

int get_M(int st)
{
    if (st == 0) return 1%d;
    return ((get_M(--st)%d)*(10%d))%d;
}

bool proc()
{
    string ss = s;
    //cout << ss << "\n";
    LL sum = 0;
    for (int i = 0; i < ss.SZ; ++i)
    sum += (((ss[i]-'0')%d)*(get_M(ss.SZ-i-1)%d))%d;
    sum %= d;
    if (!sum) return 1;
    return 0;
}

bool check()
{
    s.clear();
    //cout << "HEY\n";
    for (int i = 0; i < n; ++i){
        //cout << num[str[i]-'0'] << "\n";
        s += num[str[i]-'0'];
        //cout << s << "\n";
    }
    //cout << s << "\n";
    //cout << s.SZ << "\n";
    //for (int i = 0; i < s.SZ; ++i)
        //ch = s[i]+ch;
        //ch += (s[i]-'0')*pow(10, s.SZ-i-1);
    //cout << s << " " << ch << "\n";
    if (proc()) return true;
    return false;
}

void dfs_R(int ind)
{
    //cout << str << "\n";
    if (ind == n && check()){
        //cout << str << "\n";
        ans.insert(s);
        //cout << str << "\n";
    }
    else
        for (int i = 1; i <= n; i++)
            if (used[i] == 0)
            {
                used[i] = 1;
                str[ind] = (i+'0');
                dfs_R(ind+1);
                used[i] = 0;
            }
}

void in()
{
    ifstream cin("input.txt");
    cin >> n >> d;
    num.resize(n+1);
    for (int i = 1; i <= n; ++i)
        cin >> num[i];
}

void solution()
{
    for (int i = 1; i <= n; i++){
        el.push_back(i);
        str+= ' ';
    }
    dfs_R(0);
}


void out()
{
    ofstream cout("output.txt");
    cout << ans.SZ << "\n";
}

int main()
{
    in();
    solution();
    out();
    return 0;
}
